sklearn/cluster/_bisect_k_means.py

```diff
@@ -0,0 +1,542 @@
+
+    Attributes
+    ----------
+    cluster_centers_ : ndarray of shape (n_clusters, n_features)
+        Coordinates of cluster centers. If the algorithm stops before fully
+        converging (see ``tol`` and ``max_iter``), these will not be
+        consistent with ``labels_``.
+
+    labels_ : ndarray of shape (n_samples,)
+        Labels of each point.
+
+    inertia_ : float
+        Sum of squared distances of samples to their closest cluster center,
+        weighted by the sample weights if provided.
+
+    n_features_in_ : int
+        Number of features seen during :term:`fit`.
+
+    feature_names_in_ : ndarray of shape (`n_features_in_`,)
+        Names of features seen during :term:`fit`. Defined only when `X`
+        has feature names that are all strings.
+
+    See Also
+    --------
+    KMeans : Original implementation of K-Means algorithm.
+
+    Notes
+    -----
+    Bisection cannot be performed if n_cluster < 2.
+    Despite that in case when n_cluster == 1 - centroid will be created and points
+    will be assigned to it.
+
+    Also it might be inefficient when n_cluster is equal to 2, due to unnecassary
+    calculations for that case.
+
+    Examples
+    --------
+    >>> from sklearn.cluster import BisectingKMeans
+    >>> import numpy as np
+    >>> X = np.array([[1, 2], [1, 4], [1, 0],
+    ...               [10, 2], [10, 4], [10, 0],
+    ...               [10, 6], [10, 8], [10, 10]])
+    >>> bisect_means = BisectingKMeans(n_clusters=3, random_state=0).fit(X)
+    >>> bisect_means.labels_
+    array([2, 2, 2, 0, 0, 0, 1, 1, 1], dtype=int32)
+    >>> bisect_means.predict([[0, 0], [12, 3]])
+    array([2, 0], dtype=int32)
+    >>> bisect_means.cluster_centers_
+    array([[10.,  2.],
+           [10.,  8.],
+           [ 1., 2.]])
+    """
+
+    def __init__(
+        self,
+        n_clusters=8,
+        init="random",
+        n_init=1,
+        random_state=None,
+        max_iter=30,
+        verbose=0,
+        tol=1e-4,
+        copy_x=True,
+        algorithm="lloyd",
+        bisecting_strategy="biggest_inertia",
+    ):
+
+        super().__init__(
+            n_clusters=n_clusters,
+            init=init,
+            max_iter=max_iter,
+            verbose=verbose,
+            random_state=random_state,
+            tol=tol,
+            n_init=n_init,
+        )
+
+        self.copy_x = copy_x
+        self.algorithm = algorithm
+        self.bisecting_strategy = bisecting_strategy
+
+    def _check_params(self, X):
+        super()._check_params(X)
+
+        # bisecting_strategy
+        if self.bisecting_strategy not in ["biggest_inertia", "largest_cluster"]:
+            raise ValueError(
+                "Bisect Strategy must be 'biggest_inertia' or 'largest_cluster'. "
+                f"Got {self.bisecting_strategy} instead."
+            )
+
+        # Regular K-Means should do less computations when there are only
```

Comment:
I am not sure this warning is really necessary. Using clustering with 2 clusters is not the most standard usecase, and BisectingKMeans would not be terribly slower than KMeans.

I would remove it to avoid adding noise when varying the number of clusters.

Reply:
I agree. I removed the warning.